File: app/src/main/java/com/jbrunton/mymovies/app/moviedetails/MovieDetailsActivity.java
package com.jbrunton.mymovies.app.moviedetails;

import android.os.Bundle;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.jbrunton.mymovies.R;
import com.jbrunton.mymovies.app.helpers.PicassoHelper;
import com.jbrunton.mymovies.app.shared.BaseActivity;
import com.jbrunton.mymovies.app.shared.LoadingStateContext;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class MovieDetailsActivity extends BaseActivity<MovieDetailsViewModel> {

    private PicassoHelper picassoHelper = new PicassoHelper();
    @BindView(R.id.overview) TextView overview;
    @BindView(R.id.toolbar) Toolbar toolbar;
    @BindView(R.id.content) View content;
    @BindView(R.id.backdrop) ImageView backdrop;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_movie_details);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        setTitle("");
        bindErrorStateContext(new LoadingStateContext());

        ButterKnife.bind(this);

        viewModel().viewState().observe(this, this::updateView);
    }

    @Override public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                onBackPressed();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override protected MovieDetailsViewModel provideViewModel() {
        MovieDetailsViewModel.Factory factory = new MovieDetailsViewModel.Factory(
                getIntent().getExtras().getString("MOVIE_ID"), dependencies().moviesRepository());
        return getViewModel(MovieDetailsViewModel.class, factory);
    }

    @OnClick(R.id.error_try_again) public void tryAgain() {
        viewModel().retry();
    }

    private void updateView(MovieDetailsViewState viewState) {
        updateLoadingView(viewState.loadingViewState());

        setTitle(viewState.movie().title());
        overview.setText(viewState.movie().overview());

        picassoHelper.loadImage(this, backdrop, viewState.movie().backdropUrl());
    }
}

File: app/src/main/java/com/jbrunton/mymovies/moviedetails/MovieDetailsActivity.kt
package com.jbrunton.mymovies.moviedetails

import android.os.Bundle
import android.view.MenuItem
import android.view.View
import androidx.appcompat.widget.Toolbar
import com.jbrunton.entities.models.LoadingState
import com.jbrunton.mymovies.R
import com.jbrunton.mymovies.helpers.PicassoHelper
import com.jbrunton.mymovies.helpers.observe
import com.jbrunton.mymovies.movies.MovieViewState
import com.jbrunton.mymovies.shared.BaseActivity
import com.jbrunton.mymovies.shared.LoadingLayoutManager
import kotlinx.android.synthetic.main.activity_movie_details.*
import kotlinx.android.synthetic.main.content_movie_details.*
import kotlinx.android.synthetic.main.layout_loading_state.*
import org.koin.androidx.viewmodel.ext.android.viewModel
import org.koin.core.parameter.parametersOf

class MovieDetailsActivity : BaseActivity<MovieDetailsViewModel>() {
    val viewModel: MovieDetailsViewModel by viewModel { parametersOf(movieId()) }
    lateinit var loadingLayoutManager: LoadingLayoutManager
    private val picassoHelper = PicassoHelper()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_movie_details)
        val toolbar = findViewById<View>(R.id.toolbar) as Toolbar
        setSupportActionBar(toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)

        loadingLayoutManager = LoadingLayoutManager.buildFor(this, content)

        title = ""
        error_try_again.setOnClickListener { viewModel.retry() }

        viewModel.viewState.observe(this, this::updateView)
        viewModel.start()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                onBackPressed()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun movieId(): String = intent.extras["MOVIE_ID"] as String

    fun updateView(viewState: LoadingState<MovieViewState>) {
        loadingLayoutManager.updateLayout(viewState) {
            title = it.title
            overview.text = it.overview

            picassoHelper.loadImage(this, backdrop, it.backdropUrl)
        }
    }
}
